### DISCUSSION TITLE
Retrieve URL to a single file diff that was in a pull request

### DISCUSSION THREAD

**Discussion by gilzow** [community/community#55764] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

Searched the community but did not see a post related to this topic. 

**From the API, how can I retrieve the URL+fragment to the view of the diff for a single file in a pull request?** 

In the github UI, for a closed pull request, you can see the files that were changed in that pull request under the tab `Files Changed`. In that view, you can see a diff for each file that was changed. On the left hand side are links to each file that anchor to that file's diff. 

As an example (I have no affiliation with this repository and picked it at random from the [Github list of trending repositories](https://github.com/trending)), merged [pull request #686](https://github.com/bluesky-social/social-app/pull/686) for the repository [bluesky-social/social-app](https://github.com/bluesky-social/social-app), the pull request included [changes to 3 files](https://github.com/bluesky-social/social-app/pull/686/files), one of which was in [lib/hooks/useSetTitle.ts](https://github.com/bluesky-social/social-app/pull/686/files#diff-0db508b4303d674cde50c1b8a7b0ecfeb923652190ec020614a53817f59376d7).  <-- **How do I retrieve that link**?

I know how to retrieve the URL to a pull request and to the files tab directly, and I know how to retrieve a list of all files changed in a pull request, but I can not figure out how to either find or calculate the `#diff-HASH` of the fragment to point a user to the diff view of a specific changed file.

**Comment by LangLangBart** [comment, marked as answer]
> How do I retrieve that link?

- `shasum -a 256` of the `file path` seems to work, see setup below
  - If you find official documentation, please post a link
-  thanks to the hint from @nomorelint in this old thread: [Getting the anchor to files diffs in a pull request](https://web.archive.org/web/20200922103150/https://github.community/t/getting-the-anchor-to-files-diffs-in-a-pull-request/14235) (29/Jul/20)

---

### Bash script
- requires: `gh`  [cli/cli#installation](https://github.com/cli/cli#installation)
- everything else should come  standard under macOS

```bash
#!/usr/bin/env bash

# some default values from @gilzow's example
owner="${1:-"bluesky-social"}"
repo="${2:-"social-app"}"
pull_request="${3:-686}"

# The GraphQL query to collect all paths from the PullRequest
# the "--paginate/ $endCursor" setup is to fetch all pages of results, not just the first 100
# jq is built-in to boil down the response
gh api graphql --paginate --raw-field owner="$owner" --raw-field repo="$repo" --field pull_request=$pull_request --raw-field query=$'query ($owner: String!, $repo: String!, $pull_request: Int!, $endCursor: String) {
  repository(owner: $owner, name: $repo) {
    pullRequest(number: $pull_request) {
      files(first: 100, after: $endCursor) {
        totalCount
        pageInfo {
          endCursor
          hasNextPage
        }
        nodes {
          path
        }
      }
    }
  }
}' --jq '.data.repository.pullRequest.files.nodes[].path' | while IFS='' read -r line; do
	sha256_hash=$(echo -n "$line" | shasum -a 256 | cut -d ' ' -f1)
	echo "https://github.com/${owner}/${repo}/pull/${pull_request}/files#diff-${sha256_hash}"
done
```

```bash
# results
./script.sh
# https://github.com/bluesky-social/social-app/pull/686/files#diff-36de9238965be77e2516411ab3985b1b8834b813bbb1f56c81ddd22ddff6d437
# https://github.com/bluesky-social/social-app/pull/686/files#diff-0db508b4303d674cde50c1b8a7b0ecfeb923652190ec020614a53817f59376d7
# https://github.com/bluesky-social/social-app/pull/686/files#diff-1c9c0dc91519b204b48b97a3e96a142c16f87fa29109a4358f361d1e8e534091

./script.sh cli cli 12
# https://github.com/cli/cli/pull/12/files#diff-5557fd48b3af194f7671d3a298e84dcd82b3a56367fa46c33e9b0c11e52566d7
# https://github.com/cli/cli/pull/12/files#diff-32a2d594c44bfb91b129583be9109f7035c58cbe507079e562cff8cf92bf3cbf 
```

**Reply by gilzow** [reply]
> `shasum -a 256` of the file `path` seems to work, see setup below

I wondered if it wasn't something like this. Thank you so much!

**Reply by chadwcarlson** [reply]
Thanks @LangLangBart !

**Reply by kolmar** [reply]
For many use cases you can use `gh pr diff` instead of `gh api graphql` and `jq`:
```bash
gh pr diff --name-only "https://github.com/${owner}/${repo}/pull/${pull_request}" # | while read -r line; ...
```

**Reply by LangLangBart** [reply]
> For many use cases you can use `gh pr diff` instead of `gh api graphql` and `jq`

Yes, it works well _for many use cases_, and makes the script less intimidating.

I only encounter issues when the number of files in a diff exceeds **300**. For example, in **PR 8876** on
the **cli/cli** repo, `graphql` will still deliver.
